File: Server/SQLConnector.cpp
// Filename:    SQLConnector.cpp
// Author:      Joseph DeVictoria
// Date:        Feb_29_2016
// Purpose:     SQL Connector class to handle SQLite3 queries.

#include "SQLConnector.h"
#include "Utils.h"
#include <iostream>
#include <string.h>
#include <sstream>

SQLConnector::SQLConnector() {
    sqls = sqlite3_open("db/Oldentide.db", &database);
    if (sqls) {
        std::cout << "Can't open database!" << std::endl;
        // Is this needed?
        sqlite3_close(database);
    }
    else {
        std::cout << "Oldentide database opened successfully." << std::endl;
    }
}

SQLConnector::~SQLConnector() {
    sqlite3_close(database);
}

// Executes a static SQL statement
int SQLConnector::Execute(std::string cmd) {
    char *errorMessage = NULL;
    sqls = sqlite3_exec(database, cmd.c_str(), &ExecuteCallback, 0, &errorMessage);
    if (sqls != SQLITE_OK) {
        std::cout << "Could not Execute SQL query! Return Code:" << sqls << std::endl;
    }
    if (errorMessage) {
        // Print out the error message if any
        std::cout << "SQL ERROR MESSAGE: " << errorMessage << std::endl;
        // Free the error message, since it was alloced in exec()
        sqlite3_free(errorMessage);
    }
    return 0;
}

// Insert a new account record.
bool SQLConnector::InsertAccount(char *accountName, char *key, char *salt) {
    std::stringstream query;
    // Sanitize key, salt, and account name
    if (!utils::SanitizeAccountName(accountName)) {
        std::cout << "Account Name is invalid! Cannot insert account record" << std::endl;
        return false;
    }
    if (!utils::SanitizeHexString(key)) {
        std::cout << "Key is invalid! Cannot insert account record" << std::endl;
        return false;
    }
    if (!utils::SanitizeHexString(salt)) {
        std::cout << "Salt is invalid! Cannot insert account record" << std::endl;
        return false;
    }
    query << "insert into accounts (accountname, key, salt) values (";
    query << "\"" << accountName << "\",";
    query << "\"" << key << "\",";
    query << "\"" << salt << "\")";
    Execute(query.str());
    if (sqls == SQLITE_OK) {
        return true;
    }
    else {
        return false;
    }
}

// Inserts a new player into the database.
bool SQLConnector::InsertPlayer(Player p) {
    std::stringstream query;
    query << "INSERT INTO players (accountid, firstname, lastname, race, gender, face, skin, ";
    query << "profession, level, hp, maxhp, bp, maxbp, mp, maxmp, ep, maxep, strength, ";
    query << "constitution, intelligence, dexterity, axe, dagger, unarmed, hammer, polearm, ";
    query << "spear, staff, sword, archery, crossbow, sling, thrown, armor, dualweapon, shield, ";
    query << "bardic, conjuring, druidic, illusion, necromancy, sorcery, shamanic, summoning, ";
    query << "spellcraft, focus, armorsmithing, tailoring, fletching, weaponsmithing, ";
    query << "alchemy, lapidary, calligraphy, enchanting, herbalism, hunting, mining, ";
    query << "bargaining, camping, firstaid, lore, picklocks, scouting, search, stealth, ";
    query << "traps, aeolandis, hieroform, highgundis, oldpraxic, praxic, runic, head, ";
    query << "chest, arms, hands, legs, feet, cloak, necklace, ringone, ringtwo, righthand, ";
    query << "lefthand, zone, x, y, z, pitch, yaw) VALUES (";
    query << "'" << p.GetAccount() << "', ";
    query << "'" << p.GetName() << "', ";
    query << "'" << p.GetLastname() << "', ";
    query << "'" << p.GetRace() << "', ";
    query << "'" << p.GetGender() << "', ";
    query << "'" << p.GetFace() << "', ";
    query << "'" << p.GetSkin() << "', ";
    query << "'" << p.GetProfession() << "', ";
    query << p.GetLevel() << ", ";
    query << p.GetHp() << ", ";
    query << p.GetMaxHp() << ", ";
    query << p.GetBp() << ", ";
    query << p.GetMaxBp() << ", ";
    query << p.GetMp() << ", ";
    query << p.GetMaxMp() << ", ";
    query << p.GetEp() << ", ";
    query << p.GetMaxEp() << ", ";
    query << p.GetStrength() << ", ";
    query << p.GetConstitution() << ", ";
    query << p.GetIntelligence() << ", ";
    query << p.GetDexterity() << ", ";
    query << p.GetAxe() << ", ";
    query << p.GetDagger() << ", ";
    query << p.GetUnarmed() << ", ";
    query << p.GetHammer() << ", ";
    query << p.GetPolearm() << ", ";
    query << p.GetSpear() << ", ";
    query << p.GetStaff() << ", ";
    query << p.GetSword() << ", ";
    query << p.GetArchery() << ", ";
    query << p.GetCrossbow() << ", ";
    query << p.GetSling() << ", ";
    query << p.GetThrown() << ", ";
    query << p.GetArmor() << ", ";
    query << p.GetDualWeapon() << ", ";
    query << p.GetShield() << ", ";
    query << p.GetBardic() << ", ";
    query << p.GetConjuring() << ", ";
    query << p.GetDruidic() << ", ";
    query << p.GetIllusion() << ", ";
    query << p.GetNecromancy() << ", ";
    query << p.GetSorcery() << ", ";
    query << p.GetShamanic() << ", ";
    query << p.GetSummoning() << ", ";
    query << p.GetSpellcraft() << ", "; 
    query << p.GetFocus() << ", ";
    query << p.GetArmorsmithing() << ", ";
    query << p.GetTailoring() << ", ";
    query << p.GetFletching() << ", ";
    query << p.GetWeaponsmithing() << ", ";
    query << p.GetAlchemy() << ", ";
    query << p.GetLapidary() << ", ";
    query << p.GetCalligraphy() << ", ";
    query << p.GetEnchanting() << ", ";
    query << p.GetHerbalism() << ", "; 
    query << p.GetHunting() << ", ";
    query << p.GetMining() << ", "; 
    query << p.GetBargaining() << ", ";
    query << p.GetCamping() << ", "; 
    query << p.GetFirstAid() << ", "; 
    query << p.GetLore() << ", ";
    query << p.GetPickLocks() << ", "; 
    query << p.GetScouting() << ", "; 
    query << p.GetSearch() << ", ";
    query << p.GetStealth() << ", "; 
    query << p.GetTraps() << ", "; 
    query << p.GetAeolandis() << ", ";
    query << p.GetHieroform() << ", "; 
    query << p.GetHighGundis() << ", ";
    query << p.GetOldPraxic() << ", "; 
    query << p.GetPraxic() << ", "; 
    query << p.GetRunic() << ", ";
    query << "'" << p.GetHead() << "', "; 
    query << "'" << p.GetChest() << "', "; 
    query << "'" << p.GetArms() << "', ";
    query << "'" << p.GetHands() << "', "; 
    query << "'" << p.GetLegs() << "', "; 
    query << "'" << p.GetFeet() << "', ";
    query << "'" << p.GetCloak() << "', "; 
    query << "'" << p.GetNecklace() << "', "; 
    query << "'" << p.GetRingOne() << "', ";
    query << "'" << p.GetRingTwo() << "', ";
    query << "'" << p.GetRightHand() << "', ";
    query << "'" << p.GetLeftHand() << "', ";
    query << "'" << p.GetZone() << "', "; 
    query << p.GetX() << ", ";
    query << p.GetY() << ", "; 
    query << p.GetZ() << ", "; 
    query << p.GetPitch() << ", ";
    query << p.GetYaw() << ")";
    std::cout << query.str() << std::endl;
    Execute(query.str());
    if (sqls == SQLITE_OK) {
        return true;
    }
    else {
        return false;
    }
}

// Lists all the accounts
void SQLConnector::ListAccounts() {
    std::stringstream query;
    query << "select * from accounts ORDER BY accountname";
    Execute(query.str());
}

// Return the salt for the passed account
bool SQLConnector::GetAccountSalt(char *accountName, char *saltStringHex) {
    // Sanitize the account name before preceeding
    if (!utils::SanitizeAccountName(accountName)) {
        return false;
    }
    char *errorMessage = NULL;
    std::stringstream query;
    query << "select salt from accounts where accountname = \"" << accountName << "\"";
    // The fourth param is passed to the callback function as a void pointer to the first param
    sqls = sqlite3_exec(database, query.str().c_str(), &ReturnStringCallback, saltStringHex, &errorMessage);
    if (sqls != SQLITE_OK) {
        std::cout << "Could not Execute SQL query! Return Code:" << sqls << std::endl;
    }
    if (errorMessage) {
        // Print out the error message if any
        std::cout << "SQL ERROR MESSAGE: " << errorMessage << std::endl;
        // Free the error message, since it was alloced in exec()
        sqlite3_free(errorMessage);
        return false;
    }
    // Check to see if the salt was retrieved
    if (utils::SanitizeHexString(saltStringHex)) {
        return true;
    }
    else {
        return false;
    }

}

// Returns the key to the passed account
int SQLConnector::GetAccountKey(char *accountName, char *keyStringHex) {
    // Sanitize the account name before preceeding
    if (!utils::SanitizeAccountName(accountName)) {
        return 0;
    }
    char *errorMessage = NULL;
    std::stringstream query;
    query << "select key from accounts where accountname = \"" << accountName << "\"";
    //std::cout << query.str() << std::endl;
    // The fourth param is passed to the callback function as a void pointer to the first param
    sqls = sqlite3_exec(database, query.str().c_str(), &ReturnStringCallback, keyStringHex, &errorMessage);
    if (sqls != SQLITE_OK) {
        std::cout << "Could not Execute SQL query! Return Code:" << sqls << std::endl;
    }
    if (errorMessage) {
        // Print out the error message if any
        std::cout << "SQL ERROR MESSAGE: " << errorMessage << std::endl;
        // Free the error message, since it was alloced in exec()
        sqlite3_free(errorMessage);
    }
    return 1;
}

// A generic callback function to sqlite3_exec() that copies a c string in the first column
// of the returned row into stringToReturn. This function assumes that only one record will be
// in the result set, or else the return value will be the value in the last row processed.
static int ReturnStringCallback(void *stringToReturn, int argc, char **argv, char **azColName) {
    strcpy((char *)stringToReturn, argv[0]);
    return 0;
}

// This is a generic callback that simply prints to screen all the values
// of the returned row after a query. Callback adapted from https://www.sqlite.org/quickstart.html
static int ExecuteCallback(void *NotUsed, int argc, char **argv, char **azColName) {
    int i;
    for(i = 0; i < argc; i++) {
        std::cout << azColName[i] << " = " << (argv[i] ? argv[i] : "NULL");
        std::cout << " | ";
    }
    std::cout << std::endl;
    return 0;
}

File: Server/src/dedicated_server/sql_connector.go
// Filename:    sql_connector.go
// Author:      Joseph DeVictoria
// Date:        June_16_2018
// Purpose:     Database / sql abstraction for Oldentide dedicated server.

package main

import "common"
import "database/sql"
import "fmt"
import "log"

// Check if the account name is already taken in the database.
func accountExists(a string) bool {
	rows, err := db.Query("SELECT accountname FROM accounts WHERE accountname='" + a + "'")
	return foundInRows(rows, err)
}

// Check if this email already has an account associated with it in the database.
func emailExists(e string) bool {
	rows, err := db.Query("SELECT accountname FROM accounts WHERE email='" + e + "'")
	return foundInRows(rows, err)
}

func createAccount(accountname string, email string, verify_key string, hashed_key string, salt_key string) bool {
	// Prepare insert statement.
	ins, err := db.Prepare("INSERT INTO accounts(valid, banned, accountname, email, gamesession, playing, verify, hash, salt) values(?, ?, ?, ?, ?, ?, ?, ?, ?)")
	common.CheckErr(err)
	// Try to populate and execute an SQL statment.
	_, err = ins.Exec("0", "0", accountname, email, "0", "0", verify_key, hashed_key, salt_key)
	if err == nil {
		return true
	} else {
		return false
	}
}

func getAccountnameFromVerifyKey(k string) string {
	rows, _ := db.Query("SELECT accountname FROM accounts WHERE verify='" + k + "'")
	//ifErrPrintErr(err)
	var accountname string = ""
	if rows != nil {
		for rows.Next() {
			rows.Scan(&accountname)
		}
	}
	rows.Close()
	return accountname
}

func activateAccount(a string) bool {
	update, err := db.Prepare("UPDATE accounts SET valid=? WHERE accountname=?")
	_, err = update.Exec("1", a)
	if err == nil {
		return true
	} else {
		return false
	}
}

func banAccount(a string) bool {
	ban, err := db.Prepare("UPDATE accounts SET banned=? WHERE accountname=?")
	_, err = ban.Exec("1", a)
	if err == nil {
		return true
	} else {
		return false
	}
}

// Queries the db to make sure that we are generating a truly unique verify key.
func generateUniqueVerify(n int) string {
	findKey := true
	for findKey {
		verify_key := common.GenerateRandomLetters(n)
		rows, err := db.Query("SELECT accountname FROM accounts WHERE verify='" + verify_key + "'")
		//ifErrPrintErr(err)
		if !foundInRows(rows, err) {
			return verify_key
		}
	}
	return ""
}

// Queries the db to make sure that we are generating a truly unique salt.
func generateUniqueSalt(n int) string {
	findKey := true
	for findKey {
		salt_key := common.GenerateRandomLetters(n)
		rows, err := db.Query("SELECT accountname FROM accounts WHERE salt='" + salt_key + "'")
		if !foundInRows(rows, err) {
			return salt_key
		}
	}
	return ""
}

func foundInRows(rows *sql.Rows, err error) bool {
	found := false
	if err == sql.ErrNoRows {
		found = false
	} else if err != nil {
		log.Fatal(err)
	} else if rows.Next() {
		found = true
	}
	rows.Close()
	return found
}

func pullPcs() []common.Pc {
	rows, err := db.Query("Select * FROM players")
	defer rows.Close()
	var pcs []common.Pc
	for rows.Next() {
		var pc common.Pc
		err = rows.Scan(
			&pc.Id,
			&pc.Accountid,
			&pc.Firstname,
			&pc.Lastname,
			&pc.Guild,
			&pc.Race,
			&pc.Gender,
			&pc.Face,
			&pc.Skin,
			&pc.Profession,
			&pc.Alive,
			&pc.Plevel,
			&pc.Dp,
			&pc.Hp,
			&pc.Maxhp,
			&pc.Bp,
			&pc.Maxbp,
			&pc.Mp,
			&pc.Maxmp,
			&pc.Ep,
			&pc.Maxep,
			&pc.Strength,
			&pc.Constitution,
			&pc.Intelligence,
			&pc.Dexterity,
			&pc.Axe,
			&pc.Dagger,
			&pc.Unarmed,
			&pc.Hammer,
			&pc.Polearm,
			&pc.Spear,
			&pc.Staff,
			&pc.Sword,
			&pc.Archery,
			&pc.Crossbow,
			&pc.Sling,
			&pc.Thrown,
			&pc.Armor,
			&pc.Dualweapon,
			&pc.Shield,
			&pc.Bardic,
			&pc.Conjuring,
			&pc.Druidic,
			&pc.Illusion,
			&pc.Necromancy,
			&pc.Sorcery,
			&pc.Shamanic,
			&pc.Spellcraft,
			&pc.Summoning,
			&pc.Focus,
			&pc.Armorsmithing,
			&pc.Tailoring,
			&pc.Fletching,
			&pc.Weaponsmithing,
			&pc.Alchemy,
			&pc.Lapidary,
			&pc.Calligraphy,
			&pc.Enchanting,
			&pc.Herbalism,
			&pc.Hunting,
			&pc.Mining,
			&pc.Bargaining,
			&pc.Camping,
			&pc.Firstaid,
			&pc.Lore,
			&pc.Picklocks,
			&pc.Scouting,
			&pc.Search,
			&pc.Stealth,
			&pc.Traps,
			&pc.Aeolandis,
			&pc.Hieroform,
			&pc.Highgundis,
			&pc.Oldpraxic,
			&pc.Praxic,
			&pc.Runic,
			&pc.Head,
			&pc.Chest,
			&pc.Arms,
			&pc.Hands,
			&pc.Legs,
			&pc.Feet,
			&pc.Cloak,
			&pc.Righthand,
			&pc.Lefthand,
			&pc.Zone,
			&pc.X,
			&pc.Y,
			&pc.Z,
			&pc.Direction,
		)
		common.CheckErr(err)
		pcs = append(pcs, pc)
	}
	return pcs
}

func pullNpcs() []common.Npc {
	rows, err := db.Query("Select * FROM npcs")
	defer rows.Close()
	var npcs []common.Npc
	for rows.Next() {
		var npc common.Npc
		err = rows.Scan(
			&npc.Id,
			&npc.Firstname,
			&npc.Lastname,
			&npc.Guild,
			&npc.Race,
			&npc.Gender,
			&npc.Face,
			&npc.Skin,
			&npc.Profession,
			&npc.Alive,
			&npc.Nlevel,
			&npc.Hp,
			&npc.Maxhp,
			&npc.Bp,
			&npc.Maxbp,
			&npc.Mp,
			&npc.Maxmp,
			&npc.Ep,
			&npc.Maxep,
			&npc.Strength,
			&npc.Constitution,
			&npc.Intelligence,
			&npc.Dexterity,
			&npc.Head,
			&npc.Chest,
			&npc.Arms,
			&npc.Hands,
			&npc.Legs,
			&npc.Feet,
			&npc.Cloak,
			&npc.Righthand,
			&npc.Lefthand,
			&npc.Zone,
			&npc.X,
			&npc.Y,
			&npc.Z,
			&npc.Direction,
		)
		common.CheckErr(err)
		npcs = append(npcs, npc)
	}
	return npcs
}

func pullItemTemplates() []common.Item_template {
	rows, err := db.Query("Select * FROM item_templates")
	defer rows.Close()
	var item_templates []common.Item_template
	for rows.Next() {
		var item_template common.Item_template
		err = rows.Scan(
			&item_template.Id,
			&item_template.Name,
			&item_template.True_name,
			&item_template.Lore_level,
			&item_template.Item_type,
			&item_template.Slot,
			&item_template.Icon,
			&item_template.Weight,
			&item_template.Encumbrance,
			&item_template.Dyeable,
			&item_template.Stackable,
			&item_template.Stack_size,
			&item_template.Usable,
			&item_template.Equipable,
			&item_template.Base_price,
			&item_template.Strength_requirement,
			&item_template.Constitution_requirement,
			&item_template.Intelligence_requirement,
			&item_template.Dexterity_requirement,
			&item_template.Skill_type_0,
			&item_template.Skill_requirement_0,
			&item_template.Skill_type_1,
			&item_template.Skill_requirement_1,
			&item_template.Skill_type_2,
			&item_template.Skill_requirement_2,
			&item_template.Skill_type_3,
			&item_template.Skill_requirement_3,
			&item_template.Skill_type_4,
			&item_template.Skill_requirement_4,
		)
        fmt.Println(&item_template.Name)
		common.CheckErr(err)
		item_templates = append(item_templates, item_template)
	}
	return item_templates
}

func pushNpcs([]common.Npc) {
	fmt.Println("Not yet implemented")
}

func getRemainingPlayerSlots(account_name string, max_player_slots int) int {
	rows, err := db.Query("Select * FROM players INNER JOIN accounts ON players.account_id=accounts.id WHERE accountname=?", account_name)
	common.CheckErr(err)
	defer rows.Close()
	num_players := max_player_slots
	for rows.Next() {
		num_players--
	}
	return num_players
}

func playerFirstNameTaken(player_firstname string) bool {
	rows, err := db.Query("Select * FROM players WHERE firstname=?", player_firstname)
	common.CheckErr(err)
	return foundInRows(rows, err)
}

func addNewPlayer(player common.Pc) {
	// Need to add this...
	ins, err := db.Prepare("INSERT INTO players(account_id, firstname, lastname, guild, race, gender, face, skin, profession, alive, plevel, dp, hp, maxhp, bp, maxbp, mp, maxmp, ep, maxep, strength, constitution, intelligence, dexterity, axe, dagger, unarmed, hammer, polearm, spear, staff, sword, archery, crossbow, sling, thrown, armor, dualweapon, shield, bardic, conjuring, druidic, illusion, necromancy, sorcery, shamanic, spellcraft, summoning, focus, armorsmithing, tailoring, fletching, weaponsmithing, alchemy, lapidary, calligraphy, enchanting, herbalism, hunting, mining, bargaining, camping, firstaid, lore, picklocks, scouting, search, stealth, traps, aeolandis, hieroform, highgundis, oldpraxic, praxic, runic, head, chest, arms, hands, legs, feet, cloak, necklace, ringone, ringtwo, righthand, lefthand, zone, x, y, z, direction) values(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)")
	common.CheckErr(err)
	// Try to populate and execute an SQL statment.
	_, err = ins.Exec(
		player.Accountid,
		player.Firstname,
		player.Lastname,
		player.Guild,
		player.Race,
		player.Gender,
		player.Face,
		player.Skin,
		player.Profession,
		player.Alive,
		player.Plevel,
		player.Dp,
		player.Hp,
		player.Maxhp,
		player.Bp,
		player.Maxbp,
		player.Mp,
		player.Maxmp,
		player.Ep,
		player.Maxep,
		player.Strength,
		player.Constitution,
		player.Intelligence,
		player.Dexterity,
		player.Axe,
		player.Dagger,
		player.Unarmed,
		player.Hammer,
		player.Polearm,
		player.Spear,
		player.Staff,
		player.Sword,
		player.Archery,
		player.Crossbow,
		player.Sling,
		player.Thrown,
		player.Armor,
		player.Dualweapon,
		player.Shield,
		player.Bardic,
		player.Conjuring,
		player.Druidic,
		player.Illusion,
		player.Necromancy,
		player.Sorcery,
		player.Shamanic,
		player.Spellcraft,
		player.Summoning,
		player.Focus,
		player.Armorsmithing,
		player.Tailoring,
		player.Fletching,
		player.Weaponsmithing,
		player.Alchemy,
		player.Lapidary,
		player.Calligraphy,
		player.Enchanting,
		player.Herbalism,
		player.Hunting,
		player.Mining,
		player.Bargaining,
		player.Camping,
		player.Firstaid,
		player.Lore,
		player.Picklocks,
		player.Scouting,
		player.Search,
		player.Stealth,
		player.Traps,
		player.Aeolandis,
		player.Hieroform,
		player.Highgundis,
		player.Oldpraxic,
		player.Praxic,
		player.Runic,
		player.Head,
		player.Chest,
		player.Arms,
		player.Hands,
		player.Legs,
		player.Feet,
		player.Cloak,
		player.Necklace,
		player.Ringone,
		player.Ringtwo,
		player.Righthand,
		player.Lefthand,
		player.Zone,
		player.X,
		player.Y,
		player.Z,
		player.Direction,
	)
	common.CheckErr(err)
}
